First draft of component based guidelines (#604)
publiccodenet/standard: criteria/document-objectives.md

## Requirements

* The codebase MUST contain documentation of its objectives – like a mission and goal statement – that is understandable by developers and designers so that they can use or contribute to the codebase.

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Author: > The components the codebase is composed of MUST be documented by their functional isolation.

That implies a MUST in itself. Do we want to be that absolute here?

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Reviewer: Does it? I read my suggestion as "If there are components in the codebase, they MUST be documented by their functional isolation. But if there are no components, no documentation about functional isolation is needed." but perhaps I need to write that more clearly then.

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Author: I'm not sure if `function` has the same meaning as `functional`. The core of component based architecture is that you isolate certain functionality in components. Does that come through enough in this new sentence?

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Reviewer 2: I feel "function isolation" is not "plain English" enough. The main advantage of a module being documented by functional isolation would be reuse as a component in another codebase. With that in mind, what about:

```
Codebase modules SHOULD be documented in such a way as to enable reuse in other codebases. 
```

(Writing this as a SHOULD for now, as I'm not certain about a MUST yet.)

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Author: Maybe add `in other contexts` so you are triggered to document for unknown strangers.

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Reviewer 2: Indeed, thus:

```
Codebase modules SHOULD be documented in such a way as to enable reuse in other codebases in other contexts.
```

> * The components out of which the codebase SHOULD be composed MUST be documented by their functional isolation.

Author: Exactly

* Codebase documentation SHOULD clearly describe the connections between policy objectives and codebase objectives.
* The codebase MAY contain documentation of its objectives for the general public.

## Why this is important

Documenting codebase objectives:

> * provides an easy way for people to decide whether this codebase, or one of the components, is interesting for them now or in the future.

Reviewer 3: ```suggestion
* provides an easy way for people to decide whether this codebase, or one of its components, is interesting for them now or in the future.
```

* helps scope and isolate your own development.

> * improves maintanability and reusability.

Reviewer 3: ```suggestion
* improves maintainability and reusability.
```

> * channels (the improvements on the existing) functionality into logically isolated chunks.

Reviewer: The gist of this is good, but the parenthesis makes the sentence a bit complex to parse. Could this be rewritten in more plain English somehow? It is okay using several sentences instead of one if it helps.

Author: As a non-native english speaker it's hard to come up with a better wording. Maybe @ericherman can suggest something?

Reviewer: Perhaps "channels new functionality, and improvements to existing functionality, into logically isolated chunks."?

Author: If removing the parentheses makes all the difference, than just remove then.

> * clearly communicates to other stakeholders and contributors what the codebase, and the individual components it is composed from, is for.

Reviewer 3: ```suggestion
* clearly communicates to other stakeholders and contributors what the codebase, and the individual components it is composed of, is for.
```

Author: Maybe `it is` can be changed to `it's` as well?

Reviewer 3: Done!

## Management: what you need to do

* Add the organizational and business objectives to the codebase documentation, for example in the `README`.
* Watch for a manageable scope of project development.

## Developers and designers: what you need to do

* Add the technology and design objectives to the codebase documentation, for example in the `README`.

> * Design solutions into logically meaningfull and isolated chunks that can be reused into different contexes.

Reviewer 3: ```suggestion
* Design solutions into logically meaningful and isolated chunks that can be reused in different contexts.
```
